### PR TITLE
Contain instanceof checks in mapper merge methods

### PR DESCRIPTION
**Pull request by javanna**
We have a very general merge method in Mapper that takes another Mapper as argument. In practice, in many places we check the argument provided and cast it to the appropriate type. There are quite a few callers for this method, and in quite a few cases the argument is already the right one yet we go through checking and casting. This commit tries to clarify that by introducing package private merge methods in ObjectMapper, RootObjectMapper and NestedObjectMapper that take the expected argument type. These are directly usable in tests, which reduces the number of callers to the generic merge method and makes the code easier to follow.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java [hunk 1]
```diff
@@ -212,7 +212,11 @@ public ObjectMapper merge(Mapper mergeWith, MapperService.MergeReason reason, Ma
             MapperErrors.throwNestedMappingConflictError(mergeWith.name());
         }
         NestedObjectMapper mergeWithObject = (NestedObjectMapper) mergeWith;
-        var mergeResult = MergeResult.build(this, mergeWith, reason, parentBuilderContext);
+        return merge(mergeWithObject, reason, parentBuilderContext);
+    }
+
+    ObjectMapper merge(NestedObjectMapper mergeWithObject, MapperService.MergeReason reason, MapperBuilderContext parentBuilderContext) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java:218]: it is not directly visible in the PR, but these additional package private methods are now called from some tests in place of calling the public method that does the casting, no change needed in those tests.

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 2]
```diff
@@ -459,6 +459,17 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
     }
 
     public ObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
+        if (mergeWith instanceof ObjectMapper == false) {
+            MapperErrors.throwObjectMappingConflictError(mergeWith.name());
+        }
+        if (this instanceof NestedObjectMapper == false && mergeWith instanceof NestedObjectMapper) {
+            // TODO stop NestedObjectMapper extending ObjectMapper?
+            MapperErrors.throwNestedMappingConflictError(mergeWith.name());
+        }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:468]: Hopefully I was able to isolate these checks to the only place where they are really needed. We previously would perform them e.g. in RootObjectMapper, where we'd later cast the mapper to RootObjectMapper.

File: server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java [hunk 3]
```diff
@@ -199,9 +199,15 @@ protected MapperBuilderContext createChildContext(MapperBuilderContext mapperBui
 
     @Override
     public RootObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperBuilderContext parentBuilderContext) {
-        final var mergeResult = MergeResult.build(this, mergeWith, reason, parentBuilderContext);
+        if (mergeWith instanceof RootObjectMapper == false) {
+            MapperErrors.throwObjectMappingConflictError(mergeWith.name());
+        }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:204]: This is an interesting bit: I moved the cast to before the MergeResult#build call, and performed a more accurate check , given that the expectation is that the argument is a RootObjectMapper

**Comment by javanna** [comment]
There are some bits here that will cause merge conflicts with #102936 , so I will hold off merging this until #102936 is in.

**Review by benwtrent — APPROVED** [review]

**Comment by javanna** [comment]
Thanks for the review @benwtrent

**Pull request merged**